### ISSUE TITLE
Invalid HTTP response code (500)

### ISSUE THREAD

**Issue by GuardianMajor** [jgeboski/purple-facebook#11]
Using the provided Facebook library comes back with this error. Thoughts?

Came to the project from: https://developer.pidgin.im/ticket/16701

**Comment by dequis** [comment]
What were you trying to do when you got that error?

What operating system?

**Comment by fangfufu** [comment]
I occasionally got this error too. I couldn't reproduce it consistently, so I didn't open a ticket.

**Comment by jgeboski** [comment]
I need to add some more debugging output for HTTP stuff, which will better target these issues.  Then I'll be requesting debug logs!

**Comment by GuardianMajor** [comment]
@dequis nothing unusual, I added the account, put in my credentials and when it tried to connect it gave the error. I decided this time to click on re-enable to see what happens and it came back with something newer:

Error validating access token: Session does not match current stored session. This may be because the user changed the password since the time the session was created or Facebook has changed the session for security reasons. (452)

Which makes even less sense honestly. I am on a 64 bit Windows 8.1 Pro machine, if that makes any difference, which it shouldn't.

**Comment by GuardianMajor** [comment]
@jgeboski I would be happy to provide more debug logs if you let me know how you want me to capture and provide it for you. Just let me know.

**Comment by jgeboski** [comment]
Try readding the account.  This will flush all the session data.

**Comment by GuardianMajor** [comment]
@jgeboski  I have done that several times through the course of this, no luck :(

**Comment by jgeboski** [comment]
Can you email me complete [debugging](https://github.com/jgeboski/purple-facebook#debugging) output?  I was going to add some more verbosity to the log, but hopefully I can decipher the issue without the extra verbosity.

**Comment by curchunflo** [comment]
hello jgeboski, i have this error now, I can send you the debugging, please tellme what must I do and your email so I can send it to you

**Comment by jgeboski** [comment]
Please update to the least version of the plugin ([4098e875ebcb](https://github.com/jgeboski/purple-facebook/releases/tag/4098e875ebcb)), and directly email me [debugging](https://github.com/jgeboski/purple-facebook#debugging) output or a log file.

**Comment by dpadgett** [comment]
there doesn't seem to any new debugging stuff in the new version.  can u double check that u pushed it?

**Comment by GuardianMajor** [comment]
@jgeboski  sorry for the delay I was away during the weekend. I am not sure if I am missing it but I don't see the debugging. Can you provide steps in case I am missing something please. Thank you.

**Comment by GuardianMajor** [comment]
@jgeboski here is what I grabbed from the debug window, hoping it provides some insight, although I don't see much other than the same HTTP Error 500

```
(01:58:54) account: Connecting to account XXX.
(01:58:54) connection: Connecting. gc = 04BC7898
(01:58:54) http: Performing new request 04DC3830 to b-api.facebook.com.
(01:58:54) dnsquery: Performing DNS lookup for b-api.facebook.com
(01:58:54) dnsquery: IP resolved for b-api.facebook.com
(01:58:54) proxy: Attempting connection to 31.13.77.9
(01:58:54) proxy: Connecting to b-api.facebook.com:443 with no proxy
(01:58:54) proxy: Connection in progress
(01:58:54) proxy: Connecting to b-api.facebook.com:443.
(01:58:54) proxy: Connected to b-api.facebook.com:443.
(01:58:54) nss: SSL version 3.3 using 128-bit AES-GCM with 128-bit AEAD MAC
Server Auth: 256-bit ECDSA, Key Exchange: 256-bit ECDHE, Compression: NULL
Cipher Suite Name: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
(01:58:54) nss: subject=CN=*.facebook.com,O="Facebook, Inc.",L=Menlo Park,ST=CA,C=US issuer=CN=DigiCert High Assurance CA-3,OU=www.digicert.com,O=DigiCert Inc,C=US
(01:58:54) nss: partial certificate chain
(01:58:54) certificate/x509/tls_cached: Starting verify for b-api.facebook.com
(01:58:54) certificate/x509/tls_cached: Checking for cached cert...
(01:58:54) certificate/x509/tls_cached: ...Found cached cert
(01:58:54) nss/x509: Loading certificate from C:\Users\Major\AppData\Roaming\.purple\certificates\x509\tls_peers\b-api.facebook.com
(01:58:54) certificate/x509/tls_cached: Peer cert matched cached
(01:58:54) nss/x509: Exporting certificate to C:\Users\Major\AppData\Roaming\.purple\certificates\x509\tls_peers\b-api.facebook.com
(01:58:54) util: Writing file C:\Users\Major\AppData\Roaming\.purple\certificates\x509\tls_peers\b-api.facebook.com
(01:58:55) nss: Trusting CN=*.facebook.com,O="Facebook, Inc.",L=Menlo Park,ST=CA,C=US
(01:58:55) certificate: Successfully verified certificate for b-api.facebook.com
(01:58:55) http: Request 04DC3830 performed successfully.
(01:58:55) http: Performing new request 04DC3A28 to graph.facebook.com.
(01:58:55) dnsquery: Performing DNS lookup for graph.facebook.com
(01:58:55) dnsquery: IP resolved for graph.facebook.com
(01:58:55) proxy: Attempting connection to 31.13.77.6
(01:58:55) proxy: Connecting to graph.facebook.com:443 with no proxy
(01:58:55) proxy: Connection in progress
(01:58:55) proxy: Connecting to graph.facebook.com:443.
(01:58:55) proxy: Connected to graph.facebook.com:443.
(01:58:55) nss: SSL version 3.3 using 128-bit AES-GCM with 128-bit AEAD MAC
Server Auth: 256-bit ECDSA, Key Exchange: 256-bit ECDHE, Compression: NULL
Cipher Suite Name: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
(01:58:55) nss: subject=CN=*.facebook.com,O="Facebook, Inc.",L=Menlo Park,ST=CA,C=US issuer=CN=DigiCert High Assurance CA-3,OU=www.digicert.com,O=DigiCert Inc,C=US
(01:58:55) nss: partial certificate chain
(01:58:55) certificate/x509/tls_cached: Starting verify for graph.facebook.com
(01:58:55) certificate/x509/tls_cached: Checking for cached cert...
(01:58:55) certificate/x509/tls_cached: ...Found cached cert
(01:58:55) nss/x509: Loading certificate from C:\Users\Major\AppData\Roaming\.purple\certificates\x509\tls_peers\graph.facebook.com
(01:58:55) certificate/x509/tls_cached: Peer cert matched cached
(01:58:55) nss/x509: Exporting certificate to C:\Users\Major\AppData\Roaming\.purple\certificates\x509\tls_peers\graph.facebook.com
(01:58:55) util: Writing file C:\Users\Major\AppData\Roaming\.purple\certificates\x509\tls_peers\graph.facebook.com
(01:58:55) nss: Trusting CN=*.facebook.com,O="Facebook, Inc.",L=Menlo Park,ST=CA,C=US
(01:58:55) certificate: Successfully verified certificate for graph.facebook.com
(01:58:56) util: Writing file prefs.xml to directory C:\Users\Major\AppData\Roaming\.purple
(01:58:56) util: Writing file C:\Users\Major\AppData\Roaming\.purple\prefs.xml
(01:58:59) prefs: /pidgin/blist/x changed, scheduling save.
(01:58:59) prefs: /pidgin/blist/y changed, scheduling save.
(01:59:00) util: Writing file accounts.xml to directory C:\Users\Major\AppData\Roaming\.purple
(01:59:00) util: Writing file C:\Users\Major\AppData\Roaming\.purple\accounts.xml
(01:59:02) http: Request 04DC3A28 performed without success.
(01:59:02) connection: Connection error on 04BC7898 (reason: 16 description: Invalid HTTP response code (500))
(01:59:02) account: Disconnecting account XXX (05FADC90)
(01:59:02) connection: Disconnecting connection 04BC7898
(01:59:02) connection: Destroying connection 04BC7898
(01:59:04) prefs: /pidgin/debug/height changed, scheduling save.
(01:59:05) util: Writing file prefs.xml to directory C:\Users\Major\AppData\Roaming\.purple
(01:59:05) util: Writing file C:\Users\Major\AppData\Roaming\.purple\prefs.xml
(01:59:07) util: Writing file accounts.xml to directory C:\Users\Major\AppData\Roaming\.purple
(01:59:07) util: Writing file C:\Users\Major\AppData\Roaming\.purple\accounts.xml
```

**Comment by jgeboski** [comment]
@GuardianMajor I will need sensitive debugging information in order to dig deeper.  If you do not feel comfortable with sending the debug information to me, then discard the following instructions.

1. Open `Help => Debug Window` on the `Buddy List` window
2. Quit Pidgin with `Debug` window open
3. Open `Command Prompt` and run the following commands:
  1. `set PURPLE_UNSAFE_DEBUG=1`
  2. `set PURPLE_VERBOSE_DEBUG=1`
  3. `pidgin`
4. Recreate the issue with the `Debug` window open
5. Click `Save` in the `Debug` window
6. Send the log to me directly via email (jgeboski@gmail.com)

**Comment by GuardianMajor** [comment]
@jgeboski I'd be happy to send you what you need, I just didn't see an email address that you said to send it to you, so I posted the debug log here, also because it was the only thing I could get. So now that I have the email and the steps you need, I will gather the information and email it to you. Look for it in a few.

**Comment by jgeboski** [comment]
@GuardianMajor whoops! I thought it was enabled on my profile.  I have enabled it now.

**Comment by GuardianMajor** [comment]
I did everything you asked but it produces EXACTLY the same limited information as I pasted above, so not sure what sensitive information you are looking for but there is nothing there, so you want to take another look and make sure this is what you want?

**Comment by GuardianMajor** [comment]
@jgeboski  its ok, I have it now and I will use it but as soon as I get what you are looking for because at the moment there is nothing more in there that wasn't already there before. BTW, the last step pidgin produced unknown command.

**Comment by jgeboski** [comment]
@GuardianMajor could you try these [steps](https://github.com/jgeboski/purple-facebook#debug-log-on-windows)?

**Comment by GuardianMajor** [comment]
@jgeboski Ok, I am sending you the log file but taking a look it seems that this is the key issue:

> facebook:   Response Data: {"error":{"code":-3,"message":"Please reduce the amount of data you're asking for, then retry your request"}}

**Comment by jgeboski** [comment]
No need to send the log file! That's all I need to know.  Thanks!

**Comment by GuardianMajor** [comment]
No problem, you caught me just in time :+1:

**Comment by jgeboski** [comment]
Alright, so it looks like both of you are having the same problem.  @GuardianMajor how many friends do you have (roughly)?

**Comment by GuardianMajor** [comment]
give or take about 3614

**Comment by jgeboski** [comment]
Alright, that's certainly the issue, then.  I'll look into getting this resolved ASAP.

**Comment by GuardianMajor** [comment]
@jgeboski thank you very much, appreciate it. let me know if I can help. For some reason Github is not sending me notifications emails about responses, but I will check this regularly to see if you need anything else or if there is an update. Thank you.

**Comment by jgeboski** [comment]
Yeah, unfortunately, this is more of an "advanced" issue as I need to do more reversing.  As a result, I likely will not get to it until the end of the week.

**Comment by GuardianMajor** [comment]
@jgeboski OH that's quite alright, take your time. You are being a trooper working on it, so the least we can do is be patient and the process unfold. If you need any assistance, let me know. I am a developer myself and I appreciate the challenges - especially when a third party service is involved.

**Comment by curchunflo** [comment]
thank you james for your work, please let me know if I can do something for helping!!

**Comment by darkMatterSound** [comment]
I'm getting this error, too.  And I also can't connect through xmpp now.  Looks like you have all the info you need, but if you need my debug info, just let me know and I'll send it through!  Thanks for all the help, James!

**Comment by jgeboski** [comment]
@darkMatterSound Do you also have several thousand friends?

**Comment by darkMatterSound** [comment]
Yeah, I do.

**Comment by dpadgett** [comment]
i have only 447 and i get the error.

**Comment by jgeboski** [comment]
@dpadgett could you send me a [debugging log](https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs)?

**Comment by KaitoKid** [comment]
I got the error but after trying to reconnect a couple of times, it seems like it resolved itself. I have 350 friends

**Comment by jgeboski** [comment]
I have gotten it reversed and figured out.  Working on a patch now.

**Comment by darkMatterSound** [comment]
Woot!

**Comment by jgeboski** [comment]
I have managed to patch this locally, and things are working nicely.  I will fully test the patch tomorrow, and tag a new release.

**Comment by jgeboski** [comment]
This issue has been resolved as of [213498a7ab94](https://hg.pidgin.im/soc/2015/jgeboski/facebook/rev/213498a7ab94).  This patch will appear in the next release.  I will leave this issue open for now just to confirm this does in fact fix the issue.

**Comment by curchunflo** [comment]
hello james!! glad and happy you solved the issue!! how can I get the fix?? with an update? fresh install?? THANK YOU!!!

**Comment by dequis** [comment]
@curchunflo If you use windows, wait for the next [release][1] and when that's out, download the new `libfacebook.dll` and replace the old one with it.

[1]: https://github.com/jgeboski/purple-facebook/releases

**Comment by darkMatterSound** [comment]
I'm getting errors on facebook.c when trying to apply the patch on Linux.

**Comment by dequis** [comment]
@darkMatterSound do `git reset --hard` to ensure you're working with a clean copy, and use `./update.sh` instead to update to the latest hg version.

**Comment by darkMatterSound** [comment]
If I try to run ./update.sh I get this at the end of the process:

>autoreconf: running: /usr/bin/autoconf --force
>configure.ac:32: error: possibly undefined macro: AC_DISABLE_STATIC
>     If this token and others are legitimate, please use m4_pattern_allow.
>    See the Autoconf documentation.
>configure.ac:33: error: possibly undefined macro: AC_PROG_LIBTOOL
>autoreconf: /usr/bin/autoconf failed with exit status: 1

**Comment by jgeboski** [comment]
Running `./update` is not going to pull in the patch.  You will have to change the `VERSION` to say `facebook` rather than a hash.  Also, there is no guarantee that `./update` will patch cleanly.  In this case, you are on your own, or you are going to have to wait for me to pull in the update.

@darkMatterSound Install libtool.

**Comment by darkMatterSound** [comment]
It works!  Thanks a ton, mate!  You rock!

**Comment by jgeboski** [comment]
@darkMatterSound Thanks for the confirmation. :)

**Comment by darkMatterSound** [comment]
Okay, so it loads all FB contacts now, but all contacts show as being offline.  None are available to chat.

**Comment by jgeboski** [comment]
@darkMatterSound Try removing the account and adding it back to Pidgin.

**Comment by darkMatterSound** [comment]
Now this:
Failed to write data
![image](https://cloud.githubusercontent.com/assets/7034471/8763675/349a8746-2d5e-11e5-9549-b521e2e98744.png)

**Comment by curchunflo** [comment]
hi james, I am using gnu/linux, should I run ./update in the folder where I compile it?

**Comment by jgeboski** [comment]
Pull from git to update, or just download the latest release tarball.  I just pushed the latest release ([565fbae05259](https://github.com/jgeboski/purple-facebook/releases/tag/565fbae05259)).

**Comment by curchunflo** [comment]
hello james! i updated. I have the error again. It is rare, but when I updated and run the plugin, it worked, but a few minutes later I got the error 500 again :(

**Comment by jgeboski** [comment]
@curchunflo send me a [debugging log](https://github.com/jgeboski/purple-facebook/wiki/Debugging-Logs).

**Comment by GuardianMajor** [comment]
@jgeboski just wanted to let you know my friend that I loaded the most recent build, I got a 454 error but after I deleted the account and re-added it, it connected fine and even though it took a good 2 mins to actually load the "buddy" list, it did and it appears to be working. Thank you. I will keep you posted if anything comes up again.

**Comment by GuardianMajor** [comment]
@jgeboski about 10 mins into using it, it disconnected with the error message

>Failed to read fixed header

I hit reconnect and it did it eventually but not sure why it threw that error to begin with

**Comment by jgeboski** [comment]
Know "issue," just reconnect. This is Facebook closing the connection.

**Comment by GuardianMajor** [comment]
Ok, the reason I mentioned it is because after a couple of times happening, it crashes Pidgin altogether. So as stability goes, that's a problem. Ideas?

**Comment by jgeboski** [comment]
@GuardianMajor Please open a new issues a [report](https://developer.pidgin.im/wiki/TipsForBugReports#WindowsUsers).

**Comment by GuardianMajor** [comment]
@jgeboski yeah sorry, I wasn't saying it was your fault, I was just thinking out loud. Good work on getting FB going again, kudos.

**Comment by dequis** [comment]
@GuardianMajor Uhhh, you might have misinterpreted what he said. It looks like you found a crashing bug. This needs to be fixed. To get it fixed, you need to [open a separate issue][1] because this one is closed, and [follow these instructions to get details about the crash][2]

[1]: https://github.com/jgeboski/purple-facebook/issues/new
[2]: https://developer.pidgin.im/wiki/TipsForBugReports#WindowsUsers

**Comment by GuardianMajor** [comment]
@jgeboski ok my friend, will do as soon as I gather the information. In the meantime, FB account keeps getting disconnected with "failed to write data" and "failed to read fixed header" any idea what's causing that?

**Comment by jgeboski** [comment]
@GuardianMajor That is normal.  It just means the stream has been closed.  It will eventually get handled more gracefully.

I have locked this issue.  There 60 other messages, from the original issue, which makes your new issue hard to follow.  There are also five other people who are getting spammed when you continuously reply to this error with unrelated matters.
